html/browsers/browsing-the-web/back-forward-cache/resources/helper.sub.js

```diff
@@ -0,0 +1,116 @@
+  location.protocol === 'http:' ?
+  'http://{{host}}:{{ports[http][0]}}' :
+  'https://{{host}}:{{ports[https][0]}}';
+const originSameSite =
+  location.protocol === 'http:' ?
+  'http://{{host}}:{{ports[http][1]}}' :
+  'https://{{host}}:{{ports[https][1]}}';
+const originCrossSite =
+  location.protocol === 'http:' ?
+  'http://{{hosts[alt][www]}}:{{ports[http][0]}}' :
+  'https://{{hosts[alt][www]}}:{{ports[https][0]}}';
+
+const executorPath =
+  '/html/browsers/browsing-the-web/back-forward-cache/resources/executor.html?uuid=';
+
+// Asserts that the executor `target` is (or isn't, respectively)
+// restored from BFCache. These should be used in the following fashion:
+// 1. Call prepareNavigation() on the executor `target`.
+// 2. Navigate the executor to another page.
+// 3. Navigate back to the executor `target`.
+// 4. Call assert_bfcached() or assert_not_bfcached() on the main test HTML.
+async function assert_bfcached(target) {
+  const status = await getBFCachedStatus(target);
+  assert_implements_optional(status === 'BFCached', 'Should be BFCached');
+}
+
+async function assert_not_bfcached(target) {
+  const status = await getBFCachedStatus(target);
+  assert_implements_optional(status !== 'BFCached', 'Should not be BFCached');
+}
+
+async function getBFCachedStatus(target) {
+  const [loadCount, isPageshowFired] =
+    await target.execute_script(() => [window.loadCount, window.isPageshowFired]);
+  if (loadCount === 1 && isPageshowFired === true) {
+    return 'BFCached';
+  } else if (loadCount === 2 && isPageshowFired === false) {
+    return 'Not BFCached';
+  } else {
+    // This can occur for example when this is called before first navigating
+    // away (loadCount = 1, isPageshowFired = false), e.g. when
+    // 1. sending a script for navigation and then
+    // 2. calling getBFCachedStatus() without waiting for the completion of
+    //    the script on the `target` page.
+    assert_unreached(
+      `Got unexpected BFCache status: loadCount = ${loadCount}, ` +
+      `isPageshowFired = ${isPageshowFired}`);
+  }
+}
+
+// Always call `await remoteContext.execute_script(waitForPageShow);` after
+// triggering to navigation to the page, to wait for pageshow event on the
+// remote context.
+const waitForPageShow = async () => {
+  while (!window.pageShowPromise) {
+    await new Promise(resolve => setTimeout(resolve, 50));
+  }
+  await window.pageShowPromise;
+};
+
+// Run a test that navigates A->B->A:
+// 1. The page A is opened by `openFunc(url)`.
+// 2. `func` is executed on page A (if not null).
+// 3. The window is navigated to page B on `targetOrigin`.
+// 4. The window is back navigated to page A (expecting BFCached).
+//
+// Events `events` (an array of strings) are observed on page A and
+// `expectedEvents` (an array of strings) is expected to be recorded.
+// See `event-recorder.js` for event recording.
+function runEventTest(
+    openFunc, func, targetOrigin, events, expectedEvents, description) {
```

Comment:
`openFunc` and `func` is not varying. Should you remove this for now until it becomes relevant?

**Optional:**

To make easily extendable "params" to a test and make it more readable on the caller, I like to pass an object instead. This way, the caller can "name" the attribute and you can define default value in the test.

For instance:

Caller:
```js
doTest({
  openFunc: x => x,
  sourceOrigin: sameOrigin,
  destOrigin: sameOrigin,
})
```

Callee:
```js
const doTest = async params => {
  const default_params = {
    openFunc: null,
    someRandomAttribute: false
  }
  params = {...default_params, ...params} // Apply defaults;
  [...]
}
```

Reply:
Updated the PR using extendable params. It made the test much more concise. Thanks!
I keep `openFunc` and `func` (with fixing a bug of ignoring `openFunc` and renaming the latter), because the order of subsequent PRs that actually need `openFunc` and `func` is not yet decided. By putting helper methods in this PR, I can avoid rebasing issues in the subsequent PRs.